### ISSUE TITLE
Cannot connect to Coursera jupyter notebook

### ISSUE THREAD

**Issue by fabriziojpiva** [millejoh/emacs-ipython-notebook#239]
Hello everyone, as the title say, I am currently enrolled in a MOOC course from Coursera, and I have some .ipynb notebooks available for practicing my python skills. However, I am not being able to connect from Emacs to the notebook url to edit the code in real time. What I can do is to download the .ipynb file and open it up with `ein:notebooklist-open` after starting the jupyter server. But when it comes to connecting to the notebook url, I tried to:
1. Start the jupyter server with `ein:jupyter-server-start`
2. Open the notebook url with: `ein:notebooklist-open` or `ein:notebooklist-login`. The first one shows the error "wrong-type-argument number-or-marker-p ein: [warn] Null value passed to ein:get-ipython-major-version. Try to use ein:notebooklist-login first". Anyway, the latter doesn't work either.

Does anyone can help me? I am using GNU Emacs 25.3.1 (x86_64-pc-linux-gnu, GTK+ Version 2.24.25) and this is my .emacs file:

_(defvar myPackages
  '(better-defaults
    ein
    elpy ;; add the elpy package
    flycheck ;; add the flycheck package
    material-theme
    py-autopep8))
 (elpy-enable)				
(when (require 'flycheck nil t)
  (setq elpy-modules (delq 'elpy-module-flymake elpy-modules))
  (add-hook 'elpy-mode-hook 'flycheck-mode))
(require 'py-autopep8)
(add-hook 'elpy-mode-hook 'py-autopep8-enable-on-save)
(elpy-use-ipython)
(setq ein:jupyter-default-server-command "/usr/local/bin/jupyter")
(setq ein:jupyter-server-args (list "--no-browser"))_

The notebook URL I am trying to connect to has the form: https://hub.coursera-notebooks.org/user/XXXXXXXXXXXXXX/notebooks/Module%203.ipynb

**Comment by millejoh** [comment]
The error you are seeing happens because EIN is confused on what version of the Jupyter notebook it is trying to communicate with. Try execute `ein:force-ipython-version-check` then call `ein:notebooklist-login` and `ein:notebooklist-open` again.

Unfortunately `ein:jupyter-server-start` is not as robust as I would like it to be. If all were going well you should not be needing to call either `ein:notebooklist-open` or `ein:notebooklist-login.` Normally one only uses those when starting jupyter from the command line outside of Emacs, or when trying to connect to a remote-running jupyter session.

Another check is the setting for the `ein:jupyter-server-run-timeout` customizable variable. Whatever value it has now, try doubling it.

If you continue to have issues send me, if you can, dumps from the `*ein:jupyter-server*` and `*ein:log-all*` buffers

**Comment by fabriziojpiva** [comment]
Hello @millejoh thank you for replying. I cannot execute `ein:force-ipython-version-check` before starting the server, but after. Anyway, when doing it after, I get _"Null value passed to ein:get-ipython-major-version"_. Here it's the `*ein:log-all*`:

> [info] Login to http://localhost:8888 complete. Now you can open notebook list by `ein:notebooklist-open'. @#<buffer *GNU Emacs*>
[info] Opened notebook list at http://localhost:8888 with path  @#<buffer *ein:notebooklist http://localhost:8888*>
[warn] Null value passed to ein:get-ipython-major-version. @#<buffer *ein:notebooklist http://localhost:8888*>
[warn] Null value passed to ein:get-ipython-major-version. @#<buffer *ein:notebooklist http://localhost:8888*>
[warn] Null value passed to ein:get-ipython-major-version. @#<buffer *ein:notebooklist http://localhost:8888*>
[info] Failed to login to http://localhost:8888 @#<buffer *ein:notebooklist http://localhost:8888*>
[info] Failed to login to http://localhost:8888 @#<buffer *ein:notebooklist http://localhost:8888*>

**Comment by millejoh** [comment]
Aha, I see better now. I believe coursera is using jupyter hub, so the process for logging in is different. Try `ein:jupyterhub-connect`, then enter your username and password and keep your fingers crossed.

I've tested on a local jupyterhub setup, but who knows how this code works in the real world.

**Comment by fabriziojpiva** [comment]
Thanks for replying! But sadly it doesn't work. Here it's the `*Messages*` log:

_Loading ido...done
Loading server...done
[yas] Reloaded everything (snippets will load just-in-time).... [2 times]
For information about GNU Emacs and the GNU system, type C-h C-a.
Restarting server
Making completion list... [2 times]
You can run the command ‘ein:jupyterhub-connect’ with M-x e-conn RET
REQUEST [error] Error from parser ein:json-read: (json-readtable-error)
REQUEST [error] Error (parse-error) while connecting to https://hub.coursera-notebooks.org/user/qjykbqkwsiawcguaxorfgl/notebooks/Module%204.ipynb/hub/api/.
deferred error : (wrong-type-argument arrayp nil)
Quit [2 times]
Making completion list... [2 times]
You can run the command ‘ein:jupyterhub-connect’ with M-x e-conn RET
REQUEST [error] Error from parser ein:json-read: (json-readtable-error)
REQUEST [error] Error (parse-error) while connecting to https://hub.coursera-notebooks.org/user/qjykbqkwsiawcguaxorfgl/notebooks/Module%204.ipynb/hub/api/.
deferred error : (wrong-type-argument arrayp nil)
Making completion list...
You can run the command ‘ein:jupyterhub-connect’ with M-x e-conn RET
REQUEST [error] Error from parser ein:json-read: (json-readtable-error)
REQUEST [error] Error (parse-error) while connecting to https://hub.coursera-notebooks.org/user/qjykbqkwsiawcguaxorfgl/notebooks/Module%204.ipynb/hub/api/.
deferred error : (wrong-type-argument arrayp nil)_

**Comment by millejoh** [comment]
You don't need to be calling both `ein:jupyterhub-connect` and `ein:jupyter-server-start`. Only one or the other.

A small clarification - `ein:jupyter-server-start` starts a jupyter notebook server on your local machine based on the values of the variables  `ein:jupyter-default-server-command` and `ein:jupyter-default-notebook-directory`. I try to do this without blocking emacs, but it can take time for the server to start up. You may need to increase the value of `ein:jupyter-server-run-timeout`. I think I was too optimistic in setting the default for this variable and will make it a bit bigger in my next commit.

I know that [jupyterhub](https://jupyterhub.readthedocs.io/en/latest/) supports different authentication methods, including OAuth, but there is some magic involved with getting OAuth to work in EIN that escapes me at the moment.

If your coursera login is your Facebook account, then for sure EIN will not be able to log in. I would need to create a Coursera account of my own and subscribe to a course in order to troubleshoot further. We will see if time permits.

**Comment by samspills** [comment]
> I know that jupyterhub supports different authentication methods, including OAuth, but there is some magic involved with getting OAuth to work in EIN that escapes me at the moment.

Any updates on that magic, @millejoh? I'd like to connect to a jupyterhub that uses oauthenticator, but I'm totally stumped on how I might be able to do that.

**Comment by dickmao** [comment]
#368 will allow:
```
C-u M-x ein:login 
```
(`ein:login`is merely a shorthand for `ein:notebooklist-login`) which will prompt for cookie name and value.  For coursera, the relevant cookie is associated with `coursera-apps.org` and looks like `wwwzfnxauheuambxgiovuf-wkspc-sid`.

You can also do this manually by appending the proper line to your `~/.emacs.d/request/curl-cookie.jar` (search netscape cookie file format).

**Comment by objectc** [comment]
> #368 will allow:
> 
> ```
> C-u M-x ein:login 
> ```
> (`ein:login`is merely a shorthand for `ein:notebooklist-login`) which will prompt for cookie name and value. For coursera, the relevant cookie is associated with `coursera-apps.org` and looks like `wwwzfnxauheuambxgiovuf-wkspc-sid`.
> 
> You can also do this manually by appending the proper line to your `~/.emacs.d/request/curl-cookie.jar` (search netscape cookie file format).

@dickmao Does this still work on Coursera jupyter hub right now? I cannot find any cookie associated with `coursera-apps.org` and looks like `wwwzfnxauheuambxgiovuf-wkspc-sid`. I do have found some values belong to Domain `.coursera.org` and `hub.coursera-notebooks.org`. However, I'm not sure if they can be used in ein. If so, could you please provide a specific example about how to connect?

**Comment by dickmao** [comment]
> @dickmao Does this still work on Coursera jupyter hub right now? I cannot find any cookie associated with `coursera-apps.org` and looks like `wwwzfnxauheuambxgiovuf-wkspc-sid`

Indeed.  Perhaps coursera made a recent change or I was just lucky.  A month ago I tested on "Mathematics for Machine Learning: PCA" and that followed a binder-like cookie auth.  That course is now closed so I cannot repeat the test.

I just chose another course "Introduction to Data Science in Python" and the cookies I see match what you describe.  It **looks like** [jupyterhub oauth](https://github.com/jupyterhub/oauthenticator) but I'm not a web developer and I can't be sure.  In any case, EIN cannot handle that kind of auth.  Sorry.
